horovod/run/elastic/driver.py

```diff
@@ -0,0 +1,272 @@
+            addresses, secret_key, self._verbose)
+
+    def record_ready(self, host, slot):
+        self._worker_registry.record_ready(host, slot)
+
+    def world_size(self):
+        return self._world_size
+
+    def local_size(self, host):
+        return len(self._host_assignments[host])
+
+    def get_slot_info(self, host, slot):
+        return self._host_assignments[host][slot] if self.has_rank_assignment(host, slot) \
+            else hosts.INVALID_SLOT_INFO
+
+    def has_rank_assignment(self, host, slot):
+        if self._discovered_hosts.is_blacklisted(host):
+            return False
+        return host in self._host_assignments and len(self._host_assignments[host]) > slot
+
+    def get_available_hosts(self):
+        return self._discovered_hosts.get_available_hosts()
+
+    def wait_for_available_hosts(self, min_np):
+        tmout = timeout.Timeout(
+            self._start_timeout,
+            message='Timed out waiting for {{activity}}. Please check that you have '
+                    'enough resources to run at least {min_np} Horovod processes.'.format(min_np=min_np))
+
+        self._wait_hosts_cond.acquire()
+        try:
+            while not self._has_available_slots(self._discovered_hosts.count_available_slots(), min_np):
+                self._wait_hosts_cond.wait(tmout.remaining())
+                tmout.check_time_out_for('minimum number of hosts to become available')
+        finally:
+            self._wait_hosts_cond.release()
+
+    def _has_available_slots(self, slots, min_np):
+        return slots >= min_np
+
+    def _activate_hosts(self, min_np):
+        logging.info('wait for available hosts: {}'.format(min_np))
+        self.wait_for_available_hosts(min_np)
+        pending_slots = self._update_host_assignments()
+        self._worker_registry.reset(self.world_size())
+        self._start_worker_processes(pending_slots)
+
+    def _discover_hosts(self):
+        first_update = True
+        while not self._shutdown.is_set():
+            self._wait_hosts_cond.acquire()
+            try:
+                if self._discovered_hosts.update_available_hosts():
+                    self._notify_workers_host_changes()
+                    self._wait_hosts_cond.notify_all()
+            except RuntimeError as e:
+                if first_update:
+                    # Misconfiguration, fail the job immediately
+                    raise
+                # Transient error, retry until timeout
+                logging.warning(str(e))
+            finally:
+                self._wait_hosts_cond.release()
+            first_update = False
+            self._shutdown.wait(DISCOVER_HOSTS_FREQUENCY_SECS)
+
+    def _notify_workers_host_changes(self):
+        current_hosts = self._host_assignments.keys()
+        available_hosts = self.get_available_hosts()
+        if not current_hosts - available_hosts and not self._can_assign_hosts(available_hosts - current_hosts):
```

Comment:
This does not cater for the situation where a `current_host` looses slots. But workers will experience exceptions on slot failure anyway.

Reassignment of hosts ans slots is initiated by workers experiencing `HostsUpdatedInterrupt` or an `HorovodInternalError` right? So there is no code here that changes `host_assignment` before workers know it will change. Otherwise we would need to sync this condition with other code in the driver (and add a comment to link those two places).

Reply:
Good point.  I think the best way to handle this is to recompute the host assignments and do a 1:1 comparison between current assignments and proposed next assignments.  If they're not different, early return.  I've made this change.

Reply:
This feels safer.

Reply:
Yes, I'll add a comment that this should be enforced (for future modifiers).